### PR TITLE
fix: body callbacks return the whole buffered body size

### PR DESCRIPTION
**Pull request by M4tteoP**
Following https://github.com/tetratelabs/proxy-wasm-go-sdk/issues/414 discussion this PR tries to:
- Fix the emulator: providing to `ProxyOnRequestBody` and `ProxyOnResponseBody` the size of the whole body buffer, not of the individual chunk.
- Clarifies the buffering usage in `http_body` example.
- ~~Adds `http_body_chunk` example showing how to read chunk by chunk while buffering.~~ → See https://github.com/tetratelabs/proxy-wasm-go-sdk/pull/421
- ~~nit: https://github.com/tetratelabs/proxy-wasm-go-sdk/pull/410~~ → See https://github.com/tetratelabs/proxy-wasm-go-sdk/pull/419
- ~~upgrades support for latest envoy and istio/proxyv2~~ → See https://github.com/tetratelabs/proxy-wasm-go-sdk/pull/420

Tentatively closes https://github.com/tetratelabs/proxy-wasm-go-sdk/issues/414.

### CODE REVIEW

File: .github/workflows/workflow.yaml [hunk 1]
```diff
@@ -58,10 +58,12 @@ jobs:
           "envoyproxy/envoy:v1.25-latest",
           "envoyproxy/envoy:v1.26-latest",
           "envoyproxy/envoy:v1.27-latest",
+          "envoyproxy/envoy:v1.28-latest",
           "istio/proxyv2:1.16.0",
           "istio/proxyv2:1.17.0",
           "istio/proxyv2:1.18.0",
           "istio/proxyv2:1.19.0",
+          "istio/proxyv2:1.20.1",
```
Comment (mathetake) [inline .github/workflows/workflow.yaml:66]: sorry could you do this in a separate PR?

File: e2e/e2e_test.go [hunk 2]
```diff
@@ -241,9 +241,9 @@ func Test_network(t *testing.T) {
 			"upstream data received",
 			"connection complete!",
 			"remote address: 127.0.0.1:",
-			"upsteam cluster matadata location[region]=ap-northeast-1",
-			"upsteam cluster matadata location[cloud_provider]=aws",
-			"upsteam cluster matadata location[az]=ap-northeast-1a",
+			"upstream cluster metadata location[region]=ap-northeast-1",
+			"upstream cluster metadata location[cloud_provider]=aws",
+			"upstream cluster metadata location[az]=ap-northeast-1a",
```
Comment (mathetake) [inline e2e/e2e_test.go:246]: also I would appreciate it if you could avoid conflating unrelated changes together. I don't care receiving many PRs at once rather than having giant commits.

File: examples/http_body/main.go [hunk 3]
```diff
@@ -112,13 +110,12 @@ func (ctx *setBodyContext) OnHttpRequestBody(bodySize int, endOfStream bool) typ
 		return types.ActionContinue
 	}
 
-	ctx.totalRequestBodySize += bodySize
```
Comment (mathetake) [inline examples/http_body/main.go:115]: I believe this fix also should be separated from the adding example PR.

**Comment by mathetake** [comment]
As I commented, please make sure that each PR has a narrow scope rather than having multiple ones so that we can investigate the history easily later

**Comment by M4tteoP** [comment]
Thanks, done (and sorry for the hiccup)!

This PR has been restricted to the main topic: emulator and related existing `http_body` example fix proposal and discussion.

Other PRs created are:
- https://github.com/tetratelabs/proxy-wasm-go-sdk/pull/419 (unrelated)
- https://github.com/tetratelabs/proxy-wasm-go-sdk/pull/420 (would be great to have it merged before this one, just for extra testing)
- https://github.com/tetratelabs/proxy-wasm-go-sdk/pull/421 (draft status, depends on this one, still opened in order to provide more context)

File: examples/http_body/main.go [hunk 4]
```diff
@@ -69,16 +69,14 @@ type setBodyContext struct {
...
-	bufferOperation       string
+	modifyResponse  bool
+	bufferOperation string
 }
 
 // Override types.DefaultHttpContext.
 func (ctx *setBodyContext) OnHttpRequestHeaders(numHeaders int, endOfStream bool) types.Action {
 	mode, err := proxywasm.GetHttpRequestHeader("buffer-replace-at")
-	if mode == "response" {
+	if err == nil && mode == "response" {
```
Comment (mathetake) [inline examples/http_body/main.go:79]: curious what's the reason for this check?

**Review by mathetake — APPROVED** [review]
cool thanks!

File: examples/http_body/main.go (hunk 4 above)
Comment (M4tteoP) [reply examples/http_body/main.go:79]: `err` is retrieved but not used. I was in doubt about whether not even retrieve it, or to ensure it was nil before relying on `mode`
